x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java

```diff
@@ -0,0 +1,332 @@
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.action.GetDeploymentStatsAction;
+import org.elasticsearch.xpack.core.ml.action.UpdateTrainedModelDeploymentAction;
+import org.elasticsearch.xpack.core.ml.inference.assignment.AssignmentStats;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignment;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignmentMetadata;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.notifications.SystemAuditor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Periodically schedules adaptive allocations scaling. This process consists
+ * of calling the trained model stats API, processing the results, determining
+ * whether scaling should be applied, and potentially calling the trained
+ * model update API.
+ */
+public class AdaptiveAllocationsScalerService implements ClusterStateListener {
+
+    record Stats(long successCount, long pendingCount, long failedCount, double inferenceTime) {
+
+        long requestCount() {
+            return successCount + pendingCount + failedCount;
+        }
+
+        double totalInferenceTime() {
+            return successCount * inferenceTime;
+        }
+
+        Stats add(Stats value) {
+            long newSuccessCount = successCount + value.successCount;
+            long newPendingCount = pendingCount + value.pendingCount;
+            long newFailedCount = failedCount + value.failedCount;
+            double newInferenceTime = newSuccessCount > 0
+                ? (totalInferenceTime() + value.totalInferenceTime()) / newSuccessCount
+                : Double.NaN;
+            return new Stats(newSuccessCount, newPendingCount, newFailedCount, newInferenceTime);
+        }
+
+        Stats sub(Stats value) {
+            long newSuccessCount = Math.max(0, successCount - value.successCount);
+            long newPendingCount = Math.max(0, pendingCount - value.pendingCount);
+            long newFailedCount = Math.max(0, failedCount - value.failedCount);
+            double newInferenceTime = newSuccessCount > 0
+                ? (totalInferenceTime() - value.totalInferenceTime()) / newSuccessCount
+                : Double.NaN;
+            return new Stats(newSuccessCount, newPendingCount, newFailedCount, newInferenceTime);
+        }
+    }
+
+    /**
+     * The time interval between the adaptive allocations triggers.
+     */
+    private static final int DEFAULT_TIME_INTERVAL_SECONDS = 10;
+    /**
+     * The time that has to pass after scaling up, before scaling down is allowed.
+     */
+    private static final long SCALE_UP_COOLDOWN_TIME_MILLIS = TimeValue.timeValueMinutes(5).getMillis();
```

Comment:
Cloud has an equivalent notion of cool down time and will scale down according to it's own schedule. Scaling down frees resources (allocations) but the ml node may not be removed for some time. This makes sense to we just have to document the behaviour.

Reply:
added one line of comment here